Joshua s brown/documentation performance regression (#561)
parthenon-hpc-lab/parthenon: CONTRIBUTING.md

Total Test time (real) =   1.47 sec
```

#### Adding Performance Regression Tests

Regression tests can also be added for the purposes of testing the performance.

> Adding performance tests does require understanding a bit about how the ci

Reviewer: Please capitalize CI everywhere to be consistent with the rest of the file

infrastructure is setup. Currently, the performance tests are setup to run on
dedicated Power9 nodes that are provided by LANL, this is to ensure consistency
in the result. However, because the performance tests are running on LANL
machines it means that it is not straightforward to make the results viewable
to non LANL contributors. To make the results of the regression tests viewable

> on github a custom github application was developed to be used with Parthenon

Reviewer: ```suggestion
on GitHub a custom GitHub application was developed to be used with Parthenon
```

which we will call the Parthenon Performance Metrics App (PPMA).

The PPMA is series of python scripts located in the parthenon repository in
./scripts/python/packages/parthenon_performance_app. The PPMA has the following
responsibilities:

1. Submit a status check associated with the commit being analyzed to the
   Parthenon repository
2. Read in performance metrics from regression tests in a build directory

> 3. Compare the current metrics with metrics stored in the parthenon.wiki, the
>    files in the wiki should contain performance metrics from previous runs.

Reviewer: ```suggestion
3. Compare the current metrics with metrics stored in the Parthenon Wiki, the
   files in the Wiki should contain performance metrics from previous runs.
```
Please capitalize Wiki everywhere

4. Generate images showing the performance numbers.

> 5. Add images to an orphan branch in the parthenon repository (branch name is

Reviewer: ```suggestion
5. Add images to an orphan branch in the Parthenon repository (branch name is
```

And all other instances

   figures).

> 6. Add new performance metrics to the appropriate performance .json file stored

Reviewer: Should this be `performance.json` or just JSON file?

>    in the wiki. These json files are labeled with the following format:

Reviewer: ```suggestion
   in the Wiki. These JSON files are labeled with the following format:
```
Would prefer capitalized JSON everywhere (also instead of ".json" except when referring to a specific file)

   performance_metrics_current-branch.json, this means each branch should
   have its own .json performance file.
7. Create or overwrite a wiki page that contains links to the performance
   figures located in the parthenon wiki repository.
8. Upload the wiki page the images and performance metrics to the Parthenon
   repository. The wiki page will be named with the following format:
   current-branch_target-branch.md
   E.g. If you were merging hotfix branch into develop the wiki page would be

>    named: hotfix_develop.md, each pr should have it's own wiki page. For performance

Reviewer: ```suggestion
   named: hotfix_develop.md, each PR should have it's own Wiki page. For performance
```
What if we have more than one PR with the same branch names? I.e. PR X merges "hotfix" into "develop" and then later PR Y also merges "hotfix" into develop? Will the second overwrite the first? Would it be possible to add the PR number to the Wiki page name?

   tests that are not run as part of a wiki such as the develop branch is scheduled
   to run daily the current-branch and target-branch are the same: develop_develop.md.
9. Submit a status check indicating that the performance metrics are ready to
   be viewed with a link to the wiki page.

In order to add a performance test the following tasks must be completed.

1. The test must have a folder in ./tst/regression/test_suites (we will call it

>    dir_foo_performance). NOTE the test folder must have performance in name because

Reviewer: This is a bit confusing. What's the actual test name? Is it `dir_foo` or `foo`? Because you call the resulting metrics file `foo_results.txt` below, so that seems to indicate that we're talking about test `foo`. So I think it would be less confusing to just call this `foo_performance`.

   file ./scripts/darwin/build_fast.sh, there is a line

   ```Bash
       ctest --output-on-failure -R performance -E "SpaceInstances"
   ```

2. When the test is run e.g. by calling ctest the test must output the
   performance metrics in a file `foo_results.txt` which will appear in the
   build directory e.g. ./build/.
3. The main PPMA file located in
   ./scripts/python/packages/parthenon_performance_app/bin/ must be edited. The
   PPMA script works by looping over the output directories of the regression
   tests, the loop exists in the main PPMA file and should have a logic block
   similar to what is shown below:

```python
for test_dir in all_dirs:
    if not isinstance(test_dir, str):
        test_dir = str(test_dir)
    if (test_dir == "advection_performance"
        or test_dir == "advection_performance_mpi"):

        figure_url, figure_file, _ = self._createFigureURLPathAndName(
                    test_dir, current_branch, target_branch
                )

        # do something...

        if create_figures:
            figure_files_to_upload.append(figure_file)
            figure_urls.append(figure_url)

> 
> 

Reviewer: ```suggestion
```

```

Above you can see a branch for `advection_performance` and
`advection_performance_mpi`, because the logic is the same for both tests there is
a single if statement. You will likely need to add a separate branch for any
additional tests e.g. `dir_foo_performance`.

> The `_createFigureURLPathAndName` is a convenience method for generating a figure url

Reviewer: ```suggestion
The `_createFigureURLPathAndName` is a convenience method for generating a figure URL
```

> and a figure_file name. The figure_url will be of the form https://github.com/lanl/parthenon/blob/figures/figure_file.

Reviewer: ```suggestion
and a figure_file name. The `figure_url` will be of the form https://github.com/lanl/parthenon/blob/figures/figure_file.
```

> Using this method is not a requirement for instance if you had multiple different performance figures that you

Reviewer: ```suggestion
Using this method is not a requirement. For instance, if you had multiple different performance figures that you
```

> wanted to generate for a single test you could do it e.g.

Reviewer: ```suggestion
wanted to generate for a single test you could do, for example:
```

```python

if (test_dir == "dir_foo_performance"):
    jpeg_fig_name = "awesome.jpeg"
    figure_url1 = "https://github.com/lanl/parthenon/blob/figures/" + jpeg_fig_name

    png_fig_name = "awesome.png"
    figure_url2 = "https://github.com/lanl/parthenon/blob/figures/" + png_fig_name

    png_fig_name2 = "awesome2.png"
    figure_url3 = "https://github.com/lanl/parthenon/blob/figures/" + png_fig_name2

    # do something... need to actually generate the figures and read in performance metrics

    if create_figures:
        figure_files_to_upload.append(jpeg_fig_name)
        figure_urls.append(figure_url1)
        figure_files_to_upload.append(png_fig_name)
        figure_urls.append(figure_url2)
        figure_files_to_upload.append(png_fig_name2)
        figure_urls.append(figure_url3)
```

> 4. At this point we know the app knows the urls of the figures and we know the

Reviewer: ```suggestion
4. At this point we know the app knows the URLs of the figures and we know the
```
